### DISCUSSION TITLE
nc_var_info.cc:15:10: fatal error: 'netcdf' file not found

### DISCUSSION THREAD

**Discussion by wxbinge** [dtcenter/METplus#2662] [Installation] [answered]
I have compiled installed NetCDF-C and NetCDF-CXX to compile several packages. However, when I try to compile MET, the package cannot find NetCDF, and I am trying to figure out what I need to do differently. I am using a bash environment. Here are my setup and procedures:

**MET Version:** 11.1.1

**Compilers:** Intel OneAPI (icx and icpx)

**Issue:** Unable to find NetCDF

<img width="796" alt="image" src="https://github.com/user-attachments/assets/e1726532-0a77-43f6-93ae-a48d3b6a5719">

NetCDF directory: I've read that NetCDF-C and NetCDF-CXX must be in the same directory.

<img width="383" alt="image" src="https://github.com/user-attachments/assets/4cb53d64-c2a0-45ab-95a1-e15121bd756d">

**In the .bashrc:** export MET_NETCDF=/home/wxbinge/work/built_with_intel_2024_old_libs_version_2.1/netcdf

I've attached my config.log, config.status, and make log files.
[config.log](https://github.com/user-attachments/files/16737482/config.log)
[config.status.txt](https://github.com/user-attachments/files/16737484/config.status.txt)
[make.log](https://github.com/user-attachments/files/16737524/make.log)

**Additional Notes:** I've also tried including the MET_NETCDFINC and MET_NETCDFLIB paths in the bashrc file, but I am still getting the same error. I've tried to use the CPPFLAGS when setting up the configuration. 

Any assistance would be greatly appreciated.

**Comment by wxbinge** [comment]
It seems the netCDF libraries I am using are too old to compile with MET v11.1.1. 

I used some newer netCDF builds on my system and got past the error. However, I am now faced with this error:

/usr/bin/ld: /home/wxbinge/work/built_with_intel_2024_old_libs_version_2.1/netcdf/lib/libnetcdf_c++4.so: undefined reference to `nc_inq_var_filter'
/usr/bin/ld: /home/wxbinge/work/built_with_intel_2024_old_libs_version_2.1/netcdf/lib/libnetcdf_c++4.so: undefined reference to `nc_def_var_filter'
icpx: error: linker command failed with exit code 1 (use -v to see invocation)

Has MET been tested with Intel OneAPI compilers (icx and icpx)?

**Reply by JohnHalleyGotway** [reply]
@wxbinge I see you're having difficulty installing MET version 11.1.1, specifically related to the external NetCDF library.

You created a GitHub issue in the [MET](https://github.com/dtcenter/MET) repository. However, I've moved this over to the [METplus](https://github.com/dtcenter/METplus) repository... and converted it to a GitHub discussion instead.

MET is the statistical engine of the METplus system, but we provide support for all METplus components (including MET) via the [METplus Discussions](https://github.com/dtcenter/METplus/discussions) forum.

I'm going to ping @jprestop on this discussion and ask her to respond. She handles the majority of the installations of METplus on a variety of platform and is best suited to answer questions about compiler and dependent library versions.

**Reply by wxbinge** [reply]
Thank you for moving the thread to the correct location.

**Comment by jprestop** [comment, marked as answer]
Hi @wxbinge.

Thank you for all of the information and the files you included here. Official support in MET for the Intel oneAPI compilers was NOT included in our latest official release, MET-11.1.1, but will be included in our upcoming MET-12.0.0 official release scheduled for December this year.  We found that many of the libraries we distributed with our tar file package needed to be updated in order to compile with the oneAPI compilers, so we'll be offering a new set of tar file of dependencies with our upcoming release.  

Do you have access to the older Intel compilers or the GNU compilers for compiling MET-11.1.1?  Are you familiar with our compilation script and tar files that we provide?  If not, please take a look at our documentation under "Sample Script for Compiling External Libraries and MET" on our [MET Download](https://dtcenter.org/community-code/model-evaluation-tools-met/download) page if you're interested in using it. Please note, however, that this documentation will be moved to our User's Guide for the MET-12.0.0 release.

Regarding your attempted compilation, in your config.log file, I see that LDFLAGS contains the following:

- -L/home/wxbinge/work/built_with_intel_2024_old_libs_version_2.1/netcdf_c/v4.3.3.1/lib 
- -L/home/wxbinge/work/built_with_intel_2024_old_libs_version_2.1/netcdf_cxx/v4.2/lib

but it does not contain:

- -L/home/wxbinge/work/built_with_intel_2024_old_libs_version_2.1/netcdf/lib

where it appears the files are now located. 

Please let us know if you have any further questions.

**Reply by wxbinge** [reply]
Hey @jprestop,

Thank you for getting back to me. After building with all the latest dependencies, I successfully compiled MET v11.1.1 with the last Intel OneAPI compilers (2024.1). I am putting my steps in a separate post.

**Reply by jprestop** [reply]
Hi @wxbinge.  Thank you for letting us know that after building with more recent versions of the dependencies you were able to successful compile METv11.1.1 with the Intel OneAPI compilers (2024.1). We appreciate your detailed follow up post.  I'll go ahead and select an answer and mark this conversation as locked to prevent future posts here. Please feel free to start a new discussion with any new questions.

**Comment by wxbinge** [comment]
Compiling MET v11.1.1 with Intel OneAPI (2024.1)

Compilers: ICX and ICPX

# ================================================================== #
# USE THESE SETTING WHEN INSTALLING MET v11.1.1 or newer with Intel OneAPI compilers
#
# ================================================================== #

# ----------------- #
# >>>> MODULES <<<< #
# ----------------- #

# ZLIB
export vZLIB="v1.3.1"
export ZLIB=${UTIL}/zlib/$vZLIB

# LIBPNG
export vLIBPNG="v1.6.43"
export LIBPNG=${UTIL}/libpng/$vLIBPNG

# JPEG
export vJPEG="v9f"
export JPEG=${UTIL}/jpeg/$vJPEG

# JASPER
export vJASPER="v4.2.4"
export JASPER=${UTIL}/jasper/$vJASPER

# SZIP
export vSZIP="v2.1.1"
export SZIP=${UTIL}/szip/$vSZIP

# ZSTD
export vZSTD="v1.5.6"
export ZSTD=${UTIL}/zstd/$vZSTD

# GSLIB
export vGSL="v2.8"
export GSL_PATH=${UTIL}/gslib/$vGSL

# ***************** START NCEPLIBS ***************** #
#
# NCEP BUFRLIB
export vBUFRLIB="v12.0.1"
export BUFRLIB=${UTIL}/nceplibs/bufrlib/$vBUFRLIB

# NCEP BACIO
export vBACIO="v2.6.0"
export BACIO=${UTIL}/nceplibs/bacio/$vBACIO

# NCEP W3EMC
export vW3EMC="v2.12.0"
export W3EMC=${UTIL}/nceplibs/w3emc/$vW3EMC

# NCEP G2
export vG2="v3.5.1"
export G2_PATH=${UTIL}/nceplibs/g2/$vG2

# NCEP G2C
export vG2C="v1.9.0"
export G2C_PATH=${UTIL}/nceplibs/g2c/$vG2C

# NCEP IP
export vIP="v5.1.0"
export IP_PATH=${UTIL}/nceplibs/ip/$vIP

# NCEP SP
export vSP="v2.5.0"
export SP_PATH=${UTIL}/nceplibs/sp/$vSP

# NCEP GRIB2 LIB
export vGRIB2LIB="v1.5.0"
export GRIB2LIB=${UTIL}/nceplibs/grib2lib/$vGRIB2LIB

# ***************** END NCEPLIBS ***************** #

# CURL
export vCURL="v8.8.0"
export CURL=${UTIL}/curl/$vCURL

# HDF4
export vHDF4="v4.3.0"
export HDF4=${UTIL}/hdf4/$vHDF4

# HDF5
export vHDF5="v1.14.4"
export iHDF5=${UTIL}/hdf5/$vHDF5

# LIBXMl2
export vXML2="v2.9.12"
export XML2=${UTIL}/libxml2/$vXML2

# NETCDF-C
export vNETCDF_C="v4.9.2"
export NETCDF_C=${UTIL}/netcdf_c/$vNETCDF_C

# NETCDF-CXX
export vNETCDF_CXX="v4.3.1"
export NETCDF_CXX=${UTIL}/netcdf_cxx/$vNETCDF_CXX

# MET
export vMET="v11.1.1"
export MET_PATH=${UTIL}/met/$vMET
export MET_NETCDF=$UTIL/netcdf

# --------------------------------------------- #
# Dynamic Library Paths and Include directories
# --------------------------------------------- #
export LD_LIBRARY_PATH=$INTELHPC/2024.1:$ZLIB/lib:$LIBPNG/lib:$JPEG/lib:$JASPER/lib:$SZIP/lib:$ZSTD/lib:$CURL/lib:$GSL_PATH/lib:$XML2/lib:$HDF4/lib:$iHDF5/lib:$UTIL/netcdf/lib:$BUFRLIB/lib:$LD_LIBRARY_PATH
export PATH=$PATH:$INTELHPC/2024.1/bin:$XML2/bin
export LDFLAGS="-L$ZLIB/lib -L$LIBPNG/lib -L$JASPER/lib -L$JPEG/lib -L$SZIP/lib -L$ZSTD/lib -L$CURL/lib -L$XMl2/lib -L$HDF4/lib -L$iHDF5/lib -L$UTIL/netcdf/lib -L$GSL_PATH/lib -L$BUFRLIB/lib"
export CPPFLAGS="-I$ZLIB/include -I$LIB/include -I$JASPER/include -I$JPEG/include -I$SZIP/include -I$ZSTD/include -I$CURL/include -I$XMl2/include -I$HDF4/include -I$iHDF5/include -I$UTIL/netcdf/include -I$GSL_PATH/include"

# ====================================================== #
# CONFIGURING LIBRARIES AND DEPENDENCIES
# ====================================================== #

##############
# ZLIB
##############
./configure --prefix=$ZLIB

##############
# LIBPNG
##############
./configure --prefix=$LIBPNG

##############
# SZIP
##############
./configure --prefix=$SZIP

##############
# JPEG
##############
./configure --prefix=$JPEG

##############
# JASPER
##############
cmake -H$UTIL/SOURCE/MET/jasper-version-4.2.4 -B$UTIL/SOURCE/MET/build -DCMAKE_INSTALL_PREFIX=$JASPER

cmake --build $UTIL/SOURCE/MET/build

cd $UTIL/SOURCE/MET/build

ctest --output-on-failure

cmake --build $UTIL/SOURCE/MET/build --target install

##############
# CURL
##############
./configure --prefix=$CURL --without-ssl --with-zstd=$ZSTD

##############
# XML2
##############
./configure --prefix=$XMl2

##############
# GSL
##############
./configure --prefix=$GSL_PATH

##############
# NCEP BUFRLIB
##############
cmake -DCMAKE_INSTALL_PREFIX=$BUFRLIB ..

##############
# NCEP BACIO
##############
cmake -DCMAKE_INSTALL_PREFIX=$BACIO ..

##############
# NCEP IP
##############
cmake -DCMAKE_INSTALL_PREFIX=$IP_PATH ..

##############
# NCEP SP
##############
cmake -DCMAKE_INSTALL_PREFIX=$SP_PATH ..

##############
# NCEP W3EMC
##############
cmake .. -DCMAKE_INSTALL_PREFIX=$W3WMC -DCMAKE_PREFIX_PATH="${BACIO}"

##############
# NCEPLIB G2
##############
# Dependencies: LIBPNG, JASPER, JPEG, NCEP-BACIO, NCEP-W3EMC
cmake .. -DCMAKE_INSTALL_PREFIX=$G2_PATH -DCMAKE_PREFIX_PATH="${JPEG_LIBRARY};${JPEG_INCLUDE_DIR};${JASPERLIB};${JASPERINC};${BACIO};${W3EMC}"

##############
# NCEPLIB G2C
##############
# Dependencies: JASPER, JPEG
cmake .. -DCMAKE_INSTALL_PREFIX=$G2C_PATH -DCMAKE_PREFIX_PATH="${JPEG_LIBRARY};${JPEG_INCLUDE_DIR};${JASPERLIB};${JASPERINC}"

###############################
# NCEP GRIB2LIB AND UTILITIES
###############################
# Dependencies: LIBPNG, JASPER, JPEG, NCEP-BACIO, NCEP-IP, NCEP-SP, NCEP-W3EMC
cmake .. -DCMAKE_INSTALL_PREFIX=$GRIB2LIB -DCMAKE_PREFIX_PATH="${JPEG_LIBRARY};${JPEG_INCLUDE_DIR};${JASPERLIB};${JASPERINC};${BACIO};${IP_PATH};${SP_PATH};${W3EMC};${G2_PATH}"

##############
# HDF4
##############
./configure --prefix=$HDF4 --with-zlib=$ZLIB --with-szlib=$SZIP --with-jpeg=$JPEG --enable-shared --disable-netcdf --disable-fortran

##############
# HDF5
##############
./configure --prefix=$iHDF5 --enable-shared --enable-fortran

##############
# NETCDF-C
##############
./configure --prefix=$NETCDF_C --disable-dap --enable-shared --enable-hdf4 --enable-netcdf4

##############
# NETCDF-CXX
##############
./configure --prefix=$NETCDF_CXX

##############
# MET v11.1.1
##############
./configure --prefix=$MET_PATH CFLAGS="-qopenmp" CPPFLAGS="-I$GSL_PATH/include" LDFLAGS="-L$GSL_PATH/lib -L$BUFRLIB/lib" BUFRLIB_NAME=-lbufr_4
